### PR TITLE
Feature/detect_xpress_at_runtime

### LINKED ISSUE #698: Load xpress at runtime
Copying Simulator we could load xpress at tuntime.
This would divide by 2 the number of assets to manage.

### PR DESCRIPTION
**Pull request by a-zakir**
- Detect Xpress at Runtime
- Release one asset for each platform

### CODE REVIEW

**Comment by JasonMarechal25** [comment]
I've tried to use it. When I run a simulation (example/SmallTestFiveCandidates) with solver = xpress but xpress missing from my env I get the following output

```Launching Problem Generation
[Full Run - Problem Generation][INFO 31-10-2023 14:55:55] rename problems: false
Could not find the Xpress shared library. Looked in: [/lib/libxprs.so', '/opt/xpressmp/lib/libxprs.so', ']. Please check environment variable XPRESSDIR

Warning: false
Warning: false
Warning: false
```

With the warning repeated a lot of times
then

```[Full Run - Problem Generation][INFO 31-10-2023 14:55:57] Problem Generation ran in: 2 seconds 
[Benders][INFO 31-10-2023 14:55:57] starting Benders mpi
                     LOG_LEVEL                                                 0
                MAX_ITERATIONS                                                -1
                  ABSOLUTE_GAP                                                10
                  RELATIVE_GAP                                             1e-06
                   RELAXED_GAP                                             1e-05
              SEPARATION_PARAM                                               0.5
            MASTER_FORMULATION                                           integer
                   AGGREGATION                                                 0
                    OUTPUTROOT                                                 .
                         TRACE                                                 1
                  SLAVE_WEIGHT                                          CONSTANT
            SLAVE_WEIGHT_VALUE                                                 3
                   MASTER_NAME                                            master
                STRUCTURE_FILE                                     structure.txt
                     INPUTROOT                                                 .
                      CSV_NAME                              benders_output_trace
                   BOUND_ALPHA                                                 1
                   SOLVER_NAME                                            XPRESS
                     JSON_FILE/home/marechaljas/CLionProjects/antares-xpansion/examples/SmallTestFiveCandidates/output/20231031-1455eco-Xpansion/expansion/out.json
      LAST_ITERATION_JSON_FILE/home/marechaljas/CLionProjects/antares-xpansion/examples/SmallTestFiveCandidates/output/20231031-1455eco-Xpansion/expansion/last_iteration.json
                    TIME_LIMIT                                             1e+12
               LAST_MASTER_MPS                             master_last_iteration
                        RESUME                                                 0
             LAST_MASTER_BASIS                             master_last_basis.bss
                    BATCH_SIZE                                                 0



Warning: false
[Benders][INFO 31-10-2023 14:55:57] error: This is line 93 of file /home/marechaljas/CLionProjects/antares-xpansion/src/cpp/multisolver_interface/SolverFactory.cpp (function create_solver)
Solver 'XPRESS' not supported

--------------------------------------------------------------------------
MPI_ABORT was invoked on rank 0 in communicator MPI_COMM_WORLD
with errorcode 1.
```

If I run with default solver it works as intended. However I do get:

```
Could not find the Xpress shared library. Looked in: [/lib/libxprs.so', '/opt/xpressmp/lib/libxprs.so', ']. Please check environment variable XPRESSDIR

Warning: false
```

With the last warning repeated several times 

I note 3 points:

- The "Warning false" obscuring the output
- The warning around xpress when we didn't ask for an xpress simulation
- The error at benders step when Xpress is not present. This should be caught earlier

File: src/cpp/multisolver_interface/environment.cc [hunk 1]
```diff
@@ -0,0 +1,498 @@
...
+
+  // This is the 'assign' section.
+  xpress_dynamic_library->GetFunction(&XPRSinit, "XPRSinit");
+  xpress_dynamic_library->GetFunction(&XPRScopyprob, "XPRScopyprob");
+  xpress_dynamic_library->GetFunction(&XPRSwritebasis, "XPRSwritebasis");
+  xpress_dynamic_library->GetFunction(&XPRSreadprob, "XPRSreadprob");
+  xpress_dynamic_library->GetFunction(&XPRSreadbasis, "XPRSreadbasis");
+  xpress_dynamic_library->GetFunction(&XPRSgetrows, "XPRSgetrows");
+  xpress_dynamic_library->GetFunction(&XPRSgetindex, "XPRSgetindex");
+  xpress_dynamic_library->GetFunction(&XPRSgetnames, "XPRSgetnames");
```
Comment (flomnes) [inline src/cpp/multisolver_interface/environment.cc:152]: You can keep here only the XPRS functions that are actually called. See for example `grep XPRS`

File: src/cpp/multisolver_interface/environment.cc [hunk 2]
```diff
@@ -0,0 +1,498 @@
+// This this file is an adaptation for Antares-Xpansion needs
+//  see orginal at https://github.com/rte-france/or-tools
```
Comment (flomnes) [inline src/cpp/multisolver_interface/environment.cc]:
```suggestion
//  see orginal at https://github.com/google/or-tools
```

File: src/cpp/multisolver_interface/environment.cc [hunk 3]
```diff
@@ -0,0 +1,498 @@
+// This this file is an adaptation for Antares-Xpansion needs
```
Comment (flomnes) [inline src/cpp/multisolver_interface/environment.cc]:
```suggestion
// This this file is an adaptation of OR-Tools for Antares-Xpansion needs
```

File: src/cpp/multisolver_interface/CMakeLists.txt [hunk 4]
```diff
@@ -68,5 +73,6 @@ if(COIN_OR)
 			Coin::CoinUtils
 			Coin::Osi
 			Coin::Cbc
+		   ${CMAKE_DL_LIBS}
```
Comment (flomnes) [inline src/cpp/multisolver_interface/CMakeLists.txt:74]: I recommend creating a library to load XPRESS, and to link it against `-ldl`

Also, `-ldl` doesn't exist in Windows

File: src/cpp/multisolver_interface/SolverXpress.cpp [hunk 5]
```diff
@@ -6,6 +6,8 @@
 
 #include "StringManip.h"
 
+using namespace operations_research_Xpansion;
```
Comment (flomnes) [inline src/cpp/multisolver_interface/SolverXpress.cpp]: `operations_research` is kind of dedicated to OR-Tools. I recommend choosing a better name here

File: src/cpp/multisolver_interface/CMakeLists.txt [hunk 6]
```diff
@@ -52,12 +57,12 @@ target_include_directories (solvers
...
-		PUBLIC
-			XPRESS::XPRESS
-	)
-endif()
+#if(XPRESS)
+	# target_link_libraries (solvers
+	# 	PUBLIC
+	# 		XPRESS::XPRESS
+	# )
+#endif()
```
Comment (flomnes) [inline src/cpp/multisolver_interface/CMakeLists.txt:63]: Remove commented-out stuff

File: src/cpp/multisolver_interface/environment.cc (hunk 1 above)
Comment (a-zakir) [reply src/cpp/multisolver_interface/environment.cc:152]: I commented out the unnecessary functions and added others especially for xpansion.  I have to do it also with defines.
this needs to be cleaned

File: src/cpp/multisolver_interface/SolverXpress.cpp (hunk 5 above)
Comment (a-zakir) [reply src/cpp/multisolver_interface/SolverXpress.cpp]: Merci

File: src/cpp/multisolver_interface/SolverFactory.cpp [hunk 7]
```diff
@@ -1,34 +1,43 @@
...
-#endif
+#include "multisolver_interface/environment.h"
+
 #ifdef COIN_OR
 #include "SolverCbc.h"
 #include "SolverClp.h"
 #endif
 #include "LogUtils.h"
 #include "multisolver_interface/SolverFactory.h"
+std::vector<std::string> tmp;
```
Comment (JasonMarechal25) [inline src/cpp/multisolver_interface/SolverFactory.cpp]: Can you give a more meaningful name to this variable please?

**Review by JasonMarechal25 — CHANGES_REQUESTED** [review]

File: src/cpp/multisolver_interface/SolverFactory.cpp (hunk 7 above)
Comment (a-zakir) [reply src/cpp/multisolver_interface/SolverFactory.cpp]: done!

**Pull request closed without merging**